homeassistant/components/broadlink/config_flow.py

```diff
@@ -64,6 +65,14 @@ async def async_step_user(self, user_input=None):
                 errors["base"] = "cannot_connect"
                 err_msg = "Device not found"
 
+            except DNSLookupError as err:
```

Comment:
I'm a bit confused. Why would we need a DNS lookup?

Reply:
We need to attach the local IP address to the packet (Broadlink protocol) and we need to bind for a discovery. If we try `socket.gethostbyname(socket.gethostname())` and the user didn't add `--net=host` to runArgs in devcontainer.json he will get this error:
```python3
>>> d = blk.discover(local_ip_address='test', timeout=3)
Traceback (most recent call last):
  File "<pyshell#10>", line 1, in <module>
    d = blk.discover(local_ip_address='test', timeout=3)
  File "/home/felipe/.local/lib/python3.6/site-packages/broadlink/__init__.py", line 114, in discover
    cs.bind((local_ip_address, 0))
socket.gaierror: [Errno -2] Name or service not known
```
Which is the same error we blow if the user provides an invalid hostname as destination IP address:
```python3
>>> d = blk.discover(discover_ip_address='test', timeout=3)
Traceback (most recent call last):
  File "<pyshell#8>", line 1, in <module>
    d = blk.discover(discover_ip_address='test', timeout=3)
  File "/home/felipe/.local/lib/python3.6/site-packages/broadlink/__init__.py", line 159, in discover
    cs.sendto(packet, (discover_ip_address, 80))
socket.gaierror: [Errno -2] Name or service not known
```
So the idea is [create a separate class](https://github.com/mjg59/python-broadlink/pull/416) to deal with this exception correctly. So the output will be:
```python
>>> d = blk.discover(local_ip_address='test', timeout=3)
Traceback (most recent call last):
  File "<pyshell#10>", line 1, in <module>
    d = blk.discover(local_ip_address='test', timeout=3)
  File "/home/felipe/.local/lib/python3.6/site-packages/broadlink/__init__.py", line 114, in discover
    cs.bind((local_ip_address, 0))
broadlink.exceptions.DNSLookupError: [Errno -4013] Failed to obtain local IP address

>>> d = blk.discover(discover_ip_address='test', timeout=3)
Traceback (most recent call last):
  File "<pyshell#8>", line 1, in <module>
    d = blk.discover(discover_ip_address='test', timeout=3)
  File "/home/felipe/.local/lib/python3.6/site-packages/broadlink/__init__.py", line 159, in discover
    cs.sendto(packet, (discover_ip_address, 80))
socket.gaierror: [Errno -2] Name or service not known
```
Do you have a better idea? I took a look at `util.get_local_ip` but I'm not sure it works for VPNs.

Reply:
The option is to use our own local address lookup logic in Home Assistant and inject via local_ip_address in broadlink.discover() so we can treat both as OSErrors, separately. What do you think?

Reply:
This is fine for now.

Reply:
won't work, especially in case of multi-homed nodes. it's not just a matter of being in a container with or without a vpn. you could simply have a simple host with two network interfaces, first one being the main one with fqdn ip attached to it, second on to talk to specific restricted devices in another vlan.
this is not meant to be rude, but the logic is wrong, if you're not running a docker, how do you pass the correct ip address ? As I said in https://github.com/home-assistant/core/issues/40077#issuecomment-694921522 the integration should expose local_ip_address to the user, it could default to this docker "--net=" option if set

Reply:
@lolorc I understand the limitations of `socket.gethostbyname(socket.gethostname())` when it comes to multiple interfaces for the same host, we already [discussed this](https://github.com/mjg59/python-broadlink/pull/378) at the library. I will consider your suggestion for a future PR, but the best scenario would be user not having to worry about this. I will see what I can do when I find time to make this better. For now, this update will work for 99% of people.

Reply:
I found a way to solve this problem for everyone and in the simplest possible way. Thank you for insisting on this. Hope I'll find time to polish [this code](https://github.com/mjg59/python-broadlink/pull/427) until the next release.